jupyter_server/serverapp.py

```diff
@@ -1050,6 +1050,20 @@ def _update_pylab(self, change):
         )
         self.exit(1)
 
+    notebook_dir = Unicode(config=True,
+        help=_("DEPRECATED, use root_dir.")
+    )
+
+    @default('notebook_dir')
+    def _default_notebook_dir(self):
+        self.log.warning(_("\n    notebook_dir is deprecated, use root_dir.\n"))
+        return self.root_dir
+
+    @validate('notebook_dir')
+    def _update_notebook_dir(self, change):
+        self.log.warning(_("\n    notebook_dir is deprecated, use root_dir.\n"))
+        self.root_dir = change['value']
```

Comment:
- With this behavior, there may be an issue when setting `root_dir`, and reading notebook_dir.

- Maybe an observer on `root_dir` setting the `notebook_dir` trait through set_trait (bypassing validation) would be better.

- However, we should be careful about incompatible setting of one vs the other?

Reply:
Oops, that should be an `observe` decorator.

Reply:
Ah, you're right. If `root_dir` changes, `notebook_dir` doesn't change.

I think setting `notebook_dir` in the `root_dir` observer should fix that. Why bypass validation? They should end up having the same validation logic right?

Reply:
How about adding a 
```python
if self.notebook_dir != change['new']:
    self.notebook_dir = change['new']
```
to the `root_dir` observer, and 
```python
if self.root_dir != change['new']:
    self.root_dir = change['new']
```
to the `notebook_dir` observer?

Reply:
This ensures the traits are equal.

Reply:
I need to do some experiments. But it is not trivial.

Reply:
totally agree it's not trivial.

Reply:
I added the check above if you want to test it. Just checks `notebook_dir` and `root_dir` are equal. If not, update the other trait.